sleap_nn/inference/predictors.py

```diff
@@ -233,63 +244,64 @@ def _predict_generator(self) -> Iterator[Dict[str, np.ndarray]]:
             self._initialize_inference_model()
 
         # Loop over data batches.
-        if self.provider == "LabelsReader":
-            for ex in self.pipeline:
+        self.pipeline.start()
+        batch_size = self.preprocess_config["batch_size"]
+        done = False
+        while not done:
+            imgs = []
+            fidxs = []
+            vidxs = []
+            org_szs = []
+            instances = []
+            for _ in range(batch_size):
+                frame = self.pipeline.frame_buffer.get()
+                if frame["image"] is None:
+                    done = True
+                    break
+                imgs.append(frame["image"].unsqueeze(dim=0))
+                fidxs.append(frame["frame_idx"])
+                vidxs.append(frame["video_idx"])
+                org_szs.append(frame["orig_size"].unsqueeze(dim=0))
+                if self.instances_key:
+                    instances.append(frame["instances"].unsqueeze(dim=0))
+            if imgs:
+                imgs = torch.concatenate(imgs, dim=0)
+                fidxs = torch.tensor(fidxs, dtype=torch.int32)
+                vidxs = torch.tensor(vidxs, dtype=torch.int32)
+                org_szs = torch.concatenate(org_szs, dim=0)
+                if self.instances_key:
+                    instances = torch.concatenate(instances, dim=0)
+                ex = {
+                    "image": imgs,
+                    "frame_idx": fidxs,
+                    "video_idx": vidxs,
+                    "orig_size": org_szs,
+                }
+                if self.instances_key:
+                    ex["instances"] = instances
+                ex["image"] = apply_normalization(ex["image"])
+                if self.preprocess_config["is_rgb"]:
+                    ex["image"] = convert_to_rgb(ex["image"])
+                else:
+                    ex["image"] = convert_to_grayscale(ex["image"])
+                if self.preprocess:
+                    scale = self.preprocess_config["scale"]
+                    if scale != 1.0:
+                        if self.instances_key:
+                            ex["image"], ex["instances"] = apply_resizer(
+                                ex["image"], ex["instances"]
+                            )
+                        else:
+                            ex["image"] = resize_image(ex["image"], scale)
+                    ex["image"] = apply_pad_to_stride(
+                        ex["image"], self.preprocess_config["max_stride"]
+                    )
```

Comment:
Move this to a separate `preprocess()` method and note in the docstring that a TODO is to move this into the inference model itself so these ops are exportable.